refactor (#326)
KxSystems/docs: docs/kb/logging.md

---
# Using log files: logging, recovery and replication

## Overview

> Software or hardware problems can cause a kdb+ server process to fail, possibly resulting in loss of data not saved to disk at the time of the failure. A kdb+ server can use logging of updates to avoid data loss when failures occur.

Reviewer: ```suggestion
Software or hardware problems can cause a kdb+ server process to fail, possibly resulting in loss of data if not saved to disk at the time of the failure. A kdb+ server can use logging updates to avoid data loss when failures occur.
```

q)
```

A listing of the current directory gives:

```bash
q)\ls
"qtest.log"
```

The system command `\l` can be used to roll the log file. 

> The current log file will be renamed with the `qdb` extension and a new log file will be created.

Reviewer: ```suggestion
The current log file is renamed with the `qdb` extension and a new log file is created.
```

this reads the data file (`.qdb`), log file, and the q script file `logTest.q`, if present. If any of the three files exists (`.q`, `.qdb`, and `.log`), they should all be in the same directory.

### Logging options

The [`-l`](../basics/cmdline.md#-l-log-updates) option is recommended if you trust (or duplicate) the machine where the server is running. The [`-L`](../basics/cmdline.md#-l-log-sync) option involves an actual disk write (assuming hardware write-cache is disabled).

Another option is to use no logging. This is used with test, read-only, read-mostly, trusted, duplicated or cache databases.

### Errors and rollbacks

> If either message handler ([`.z.pg`](../ref/dotz.md#zpg-get) or [`.z.ps`](../ref/dotz.md#zps-set)) throws any error, and the state was changed during that message processing, this will cause a rollback.

Reviewer: ```suggestion
If either message handler ([`.z.pg`](../ref/dotz.md#zpg-get), or [`.z.ps`](../ref/dotz.md#zps-set)), throws any error and the state was changed during that message processing, this initiates a rollback.
```

On start-up, the replicating process connects to the logging process, gets the log filename and record count, opens the log file, plays back that count of records from the log file, and continues to receive updates via [TCP/IP](../basics/ipc.md). Each record is executed via `value`.

If the replicating process loses its connection to the logging process, you can detect that with [`.z.pc`](../ref/dotz.md#zpc-close). To resubscribe to the logging process, restart the replicating process.

Currently, only a single replicating process can subscribe to the primary process. If another kdb+ process attempts to replicate from the primary, the previous replicating process will no longer receive updates. If you need multiple replicating processes, you might like to consider [kdb+tick](../learn/startingkdb/tick.md).

## Manual handling

### Overview

Function calls and the contents of the their parameters can be recorded to a log file, which can then be replayed by a process. This is often used for data recovery.

> This technique is often used to allow a developer more control over log file naming conventions, what to log, log file locations and the ability to add logic around the log file lifecycle.

Reviewer: ```suggestion
This technique allows more control over actions like log file naming conventions, what to log, log file locations and the ability to add logic around the log file lifecycle.
```

> ### Creating a log file

Reviewer: ```suggestion
### Create a log file
```

> 

Reviewer: To create a log file, do the following.

> A log file can be initialized using [`set`](../ref/get.md#set).

Reviewer: ```suggestion
1. Initialize a log file by using [`set`](../ref/get.md#set).
```

```q
q)logfile:hsym `$"qlog";
q)logfile set ();
q)logfilehandle:hopen logfile;
```

!!!note "`logfile set ();` is equivalent to `.[logfile;();:;()];`"

> A check for a pre-existing log file can be achieved using [`get`](../ref/get.md#get). 
> The script can now be written to initialize a new log file if it does not exist, otherwise it will opened for appending.

Reviewer: ```suggestion
1. Check for pre-existing log files by using [`get`](../ref/get.md#get). 
1. If a new log file does not exist, the script can now be written to initialize it, otherwise it is opened for appending.
```

```q
q)logfile:hsym `$"qlog";
q)if[not type key logfile;logfile set()]
q)logfilehandle:hopen logfile;
```

> Close the log file when finished logging any messages.

Reviewer: ```suggestion
1. Close the log file when you are finished logging any messages.
```

```q
q)hclose logfilehandle
```

> ### Log writting

Reviewer: ```suggestion
### Log writing
```

> To record events/messages to a log, append a list consisting of a function name followed by any parameters used. 

Reviewer: ```suggestion
To record events and messages to a log, you must append a list consisting of a function name, followed by any parameters used. 
```

A tickerplant uses this concept to record all messages sent to its clients so they can use the log to recover. It records calling a function `upd` passing the parameters of a table name and the table content to append.

> For example, calling a function called `upd` with two parameters, x and y, can be recorded to a file file as follows:

Reviewer: ```suggestion
For example, calling a function called `upd` with two parameters, x and y, can be recorded to a file as follows:
```

```q
q)logfilehandle enlist (`upd;x;y)
```

> When a kdb+ process plays back this log, a function called `upd` will be called with the value of the two parameters. Multiple function calls can be logged also:

Reviewer: ```suggestion
When a kdb+ process plays back this log, a function called `upd` is called with the value of the two parameters. Multiple function calls can be logged also:
```

```q
q)logfilehandle ((`func1;param1);(`func2;param1;param2))
```

As [log replay](#replaying-log-files) calls [`value`](../ref/value.md) to execute, you can also log q code as a string, for example

```q
logfilehandle enlist "upd[22;33]"
```

Function calls that are used to updating data are typically written, without logging the function definition. 

> This has the disadvantage of requiring that the recoverying process defines the functions (e.g. by loading a q script) prior to replaying the log. 
> The advantages can outweight the disadvantages by allowing for bugs fixes within the function or temporarily assigning the function to a different definition prior to playback, 
> to provide despoke logic for data sourced from a log file.

Reviewer: ```suggestion
This has the disadvantage of requiring that the recovery process defines the functions (for example, by loading a q script) prior to replaying the log. 
The advantages can outweigh the disadvantages, however, by allowing for bugs fixes within the function, or temporarily assigning the function to a different definition prior to playback, to provide bespoke logic for data sourced from a log file.
```

### Log rolling

A kdb+ process can run 24/7, but a log file may only be relevant for a specific timeframe or event. 
For example, the default tickerplant creates a new log for each day. 

> The developer should close the current log and create a new one on each event. 

Reviewer: ```suggestion
You should ensure the current log is closed and a new log created on each event. 
```

A decision must be taken on whether to retain the old files or delete them, taking into account disk usage and what other processes may require them.

A naming convention should be used to aid distinction between current log files and any old log files required for retention.
The [`z`](../ref/dotz.md) namespace provides various functions for system information such as current date, time, etc. 

> The following is an example of namimg a log file after the current date:

Reviewer: ```suggestion
The below example demonstrates naming a log file after the current date:
```

26
```

### Replacing a corrupt log

It can be  more efficient to [replay from a corrupt file](##replay-from-corrupt-logs) (due to disk usage), than to directly take the good chunks from a bad log to create a new log.

The knowledge of how to create a log file, and how to replay part of a log file can be combined to convert a file that was previously giving the `badtail` error.
Note that this does not fix the corrupted section, only removes the corrupted section from the file.

The following example shows converting a `bad.log` into a `good.log` by temporarly overriding [`.z.ps`](../ref/dotz.md#zps-set) which is called for each valid chunk (as defined by [`-11!`](../basics/internal.md#-11-streaming-execute)). 
It resets `.z.ts` to the system default after processing using [`\x`](../basics/syscmds/#x-expunge).

```q
goodfile:hsym `:good.log;
goodfile set ();
goodfilehandle:hopen goodfile;

chunks:first -11!(-2;`:bad.log);

.z.ps:{goodfilehandle enlist x};
-11!(chunks;`:bad.log);
system"x .z.ps";

hclose goodfilehandle;
```

> Alternatively, generic system tools can be used e.g. unix `head` command. For example, given that ``-11!(-2;`:bad.log)`` returns 2879 bytes.

Reviewer: ```suggestion
Alternatively, generic system tools can be used like the unix `head` command. For example, given that ``-11!(-2;`:bad.log)`` returns 2879 bytes.
```
